### ISSUE TITLE
Proposal: add 'Encoding' object

### ISSUE THREAD

**Issue by connollyst** [numenta/htm.java#161]
Hi @cogmission, looking through the code I see room to make the code more OO by adding an ``Encoding`` object. This would be the output of an ``Encoder`` but can also be used internally.

In working on the ``RandomDistributedScalarEncoder`` I saw that we sometimes use an ``int[]`` of the byte **indices** to represent the encoding, and in other places we represent the encoding as an ``int[]`` of the ``1``s and ``0``s themselves. I found this made the code hard to follow at times.

I also saw that there was some duplicate code because we are using primitive a ``int[]``. If we introduce an object, we can encapsulate a lot of common logic, test it, and reuse it. For example..

    Encoding x = encoder.encode(a);
    int on = x.getOnBits();
    int[] indices = x.getOnIndices();
    int[] binary = x.getOnBytes(); // byte[] would be more efficient here, another conversation

    Encoding y = encoder.encode(b);
    int overlap = x.countOverlap(b);

Initial gut reaction might be a concern for the memory/performance impact of newing Objects. However, by encapsulating it we can optimize the stored data and the variables initialized (it should be immutable).

I also think it'll make it easier to write efficient code as we won't convert between data representations unnecessarily.

@cogmission, you're most familiar with the code base, what do you think?
@anubhavchaturvedi, after working on the ``RandomDistributedScalarEncoder``, what do you think?

Cheers!
Sean

PS: is an SDR a subset of an ``Encoding``? How are SDRs represented?

**Comment by cogmission** [comment]
Sean,

Have you seen:
https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/encoders/EncoderResult.java
 ?

There is also DecodeResult. Are these what you are speaking of?

David

On Wed, Jan 21, 2015 at 5:41 AM, Sean Connolly <notifications@github.com>
wrote:

> Hi @cogmission <https://github.com/cogmission>, looking through the code
> I see room to make the code more OO by adding an Encoding object. This
> would be the output of an Encoder but can also be used internally.
>
> In working on the RandomDistributedScalarEncoder I saw that we sometimes
> use an int[] of the byte *indices* to represent the encoding, and in
> other places we represent the encoding as an int[] of the 1s and 0s
> themselves. I found this made the code hard to follow at times.
>
> I also saw that there was some duplicate code because we are using
> primitive a int[]. If we introduce an object, we can encapsulate a lot of
> common logic, test it, and reuse it. For example..
>
> Encoding x = encoder.encode(a);
> int on = x.getOnBits();
> int[] indices = x.getOnIndices();
> int[] binary = x.getOnBytes(); // byte[] would be more efficient here, another conversation
>
> Encoding y = encoder.encode(b);
> int overlap = x.countOverlap(b);
>
> Initial gut reaction might be a concern for the memory/performance impact
> of newing Objects. However, by encapsulating it we can optimize the stored
> data and the variables initialized (it should be immutable).
>
> I also think it'll make it easier to write efficient code as we won't
> convert between data representations unnecessarily.
>
> @cogmission <https://github.com/cogmission>, you're most familiar with
> the code base, what do you think?
> @anubhavchaturvedi <https://github.com/anubhavchaturvedi>, after working
> on the RandomDistributedScalarEncoder, what do you think?
>
> Cheers!
> Sean
>
> PS: is an SDR a subset of an Encoding? How are SDRs represented?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/htm.java/issues/161>.
>



-- 
*We find it hard to hear what another is saying because of how loudly "who
one is", speaks...*

**Comment by connollyst** [comment]
Yup, both actually look useful in the ``RandomDistributedScalarEncoder``.

``EncoderResult`` looks like what I was talking about, but has no functions at the moment. Perhaps it could be expanded to encapsulate things like ``getOverlap(EncoderResult other)`` and ``getOnBits()``. Then that logic could be unit tested and moved out of the ``RDSEncoder``.

``DecodeResult`` also look useful for the ``RDSEncoder`` also as (it seems) a ``range`` here is comparable to a ``bucket``, would that be correct? The use of ``MinMax`` here could simplify some of the calculations in the ``RDSEncoder``. Doing so could move logic out (eg: that nasty Java/Python rounding difference) and make it easy to unit test.

Just a suggestion. Say so if it's not necessary or not important at this time.


-Sean

**Comment by cogmission** [comment]
The only way this would be efficient is if we created something like an
SDR.java class that was created EVERYWHERE instead of an int[]. We already
create various int[]' s when creating input, feeding into the Encoder;
calling compute() on the SpatialPooler; TemporalMemory, CLAClassifier.

Creating a separate Encoding class I believe is not entirely the best way
to go. There is a discussion regarding the best representation here:
https://github.com/numenta/htm.java/issues/116 for reference.

If we could subclass a really efficient 3rd party class and name the
subclass SDR.java and use it EVERYWHERE, that might be best - but we'd have
to test to make sure that it increases efficiency across the use of the
stack of CLA entities.

David

On Wed, Jan 21, 2015 at 5:46 AM, cogmission1 . <cognitionmission@gmail.com>
wrote:

> Sean,
>
> Have you seen:
> https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/encoders/EncoderResult.java
>  ?
>
> There is also DecodeResult. Are these what you are speaking of?
>
> David
>
> On Wed, Jan 21, 2015 at 5:41 AM, Sean Connolly <notifications@github.com>
> wrote:
>
>> Hi @cogmission <https://github.com/cogmission>, looking through the code
>> I see room to make the code more OO by adding an Encoding object. This
>> would be the output of an Encoder but can also be used internally.
>>
>> In working on the RandomDistributedScalarEncoder I saw that we sometimes
>> use an int[] of the byte *indices* to represent the encoding, and in
>> other places we represent the encoding as an int[] of the 1s and 0s
>> themselves. I found this made the code hard to follow at times.
>>
>> I also saw that there was some duplicate code because we are using
>> primitive a int[]. If we introduce an object, we can encapsulate a lot
>> of common logic, test it, and reuse it. For example..
>>
>> Encoding x = encoder.encode(a);
>> int on = x.getOnBits();
>> int[] indices = x.getOnIndices();
>> int[] binary = x.getOnBytes(); // byte[] would be more efficient here, another conversation
>>
>> Encoding y = encoder.encode(b);
>> int overlap = x.countOverlap(b);
>>
>> Initial gut reaction might be a concern for the memory/performance impact
>> of newing Objects. However, by encapsulating it we can optimize the stored
>> data and the variables initialized (it should be immutable).
>>
>> I also think it'll make it easier to write efficient code as we won't
>> convert between data representations unnecessarily.
>>
>> @cogmission <https://github.com/cogmission>, you're most familiar with
>> the code base, what do you think?
>> @anubhavchaturvedi <https://github.com/anubhavchaturvedi>, after working
>> on the RandomDistributedScalarEncoder, what do you think?
>>
>> Cheers!
>> Sean
>>
>> PS: is an SDR a subset of an Encoding? How are SDRs represented?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/numenta/htm.java/issues/161>.
>>
>
>
>
> --
> *We find it hard to hear what another is saying because of how loudly "who
> one is", speaks...*
>



-- 
*We find it hard to hear what another is saying because of how loudly "who
one is", speaks...*

**Comment by cogmission** [comment]
Let's keep discussing this. My intuition is not to introduce a new type on
every return value from a given Encoder. Eventually, we are going to be
optimizing the performance - and I'm wary of creating a bottleneck, but
let's keep discussing this.

On Wed, Jan 21, 2015 at 6:01 AM, cogmission1 . <cognitionmission@gmail.com>
wrote:

> The only way this would be efficient is if we created something like an
> SDR.java class that was created EVERYWHERE instead of an int[]. We already
> create various int[]' s when creating input, feeding into the Encoder;
> calling compute() on the SpatialPooler; TemporalMemory, CLAClassifier.
>
> Creating a separate Encoding class I believe is not entirely the best way
> to go. There is a discussion regarding the best representation here:
> https://github.com/numenta/htm.java/issues/116 for reference.
>
> If we could subclass a really efficient 3rd party class and name the
> subclass SDR.java and use it EVERYWHERE, that might be best - but we'd have
> to test to make sure that it increases efficiency across the use of the
> stack of CLA entities.
>
> David
>
> On Wed, Jan 21, 2015 at 5:46 AM, cogmission1 . <cognitionmission@gmail.com
> > wrote:
>
>> Sean,
>>
>> Have you seen:
>> https://github.com/numenta/htm.java/blob/master/src/main/java/org/numenta/nupic/encoders/EncoderResult.java
>>  ?
>>
>> There is also DecodeResult. Are these what you are speaking of?
>>
>> David
>>
>> On Wed, Jan 21, 2015 at 5:41 AM, Sean Connolly <notifications@github.com>
>> wrote:
>>
>>> Hi @cogmission <https://github.com/cogmission>, looking through the
>>> code I see room to make the code more OO by adding an Encoding object.
>>> This would be the output of an Encoder but can also be used internally.
>>>
>>> In working on the RandomDistributedScalarEncoder I saw that we
>>> sometimes use an int[] of the byte *indices* to represent the encoding,
>>> and in other places we represent the encoding as an int[] of the 1s and
>>> 0s themselves. I found this made the code hard to follow at times.
>>>
>>> I also saw that there was some duplicate code because we are using
>>> primitive a int[]. If we introduce an object, we can encapsulate a lot
>>> of common logic, test it, and reuse it. For example..
>>>
>>> Encoding x = encoder.encode(a);
>>> int on = x.getOnBits();
>>> int[] indices = x.getOnIndices();
>>> int[] binary = x.getOnBytes(); // byte[] would be more efficient here, another conversation
>>>
>>> Encoding y = encoder.encode(b);
>>> int overlap = x.countOverlap(b);
>>>
>>> Initial gut reaction might be a concern for the memory/performance
>>> impact of newing Objects. However, by encapsulating it we can optimize the
>>> stored data and the variables initialized (it should be immutable).
>>>
>>> I also think it'll make it easier to write efficient code as we won't
>>> convert between data representations unnecessarily.
>>>
>>> @cogmission <https://github.com/cogmission>, you're most familiar with
>>> the code base, what do you think?
>>> @anubhavchaturvedi <https://github.com/anubhavchaturvedi>, after
>>> working on the RandomDistributedScalarEncoder, what do you think?
>>>
>>> Cheers!
>>> Sean
>>>
>>> PS: is an SDR a subset of an Encoding? How are SDRs represented?
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/numenta/htm.java/issues/161>.
>>>
>>
>>
>>
>> --
>> *We find it hard to hear what another is saying because of how loudly
>> "who one is", speaks...*
>>
>
>
>
> --
> *We find it hard to hear what another is saying because of how loudly "who
> one is", speaks...*
>



-- 
*We find it hard to hear what another is saying because of how loudly "who
one is", speaks...*

**Comment by connollyst** [comment]
I had seen that conversation in #116 and am still up in the air on that one. I'm new to the algorithm and the code base, so can't really give a valuable opinion yet.

As for having an ``SDR`` class, I think that would be super valuable. Ideally it would just be a wrapper of an int[] but would encapsulate what that ``int[]`` means and does.

If it has no other variables, the footprint overhead would be a few bytes just for the ``Object``.

We wouldn't need to worry about instantiation costs if we keep it a simple object.
To this effect I'd shy away from the ``Tuple`` implementation we have with ``EncoderResult`` - that EAV kind of map is probably useful in places, but would be wasteful for such a core object.

Think of it like a ``String``; it's a immutable wrapper of a ``char[]`` that lets you do cool things to it.

As a side bonus: by tucking all this common logic away into one place, it would be super easy to do performance  tests between different implementations. It would be nice to have some benchmarks in place beforehand so as to be confident we'd be moving in the right direction with this change.

**Comment by cogmission** [comment]
Some more thoughts...


> To this effect I'd shy away from the Tuple implementation we have with
> EncoderResult - that EAV kind of map is probably useful in places, but
> would be wasteful for such a core object.


First, I like the names Encoding/Decode better than
EncoderResult/DecodeResult. Also, these are only returned from
topDownCompute() and decode() respectively. So the two methods mentioned
are outside of the main compute chain, (they're used to gather
meta-information or statistics for the most part) - making them ok to
elaborate on in the way Sean suggests.

I call the "main compute chain" the methods called to enter the Encoder,
SpatialPooler, TemporalMemory, XXXClassifier. I think these should remain
as highly optimized as possible.

The SDR conversation and the canonical representation format are a
different conversation because the "Encoding" class is not used in the main
compute chain, and is returned from a method that isn't crucial to the
computing pipeline. The SDR conversation can be put off for now.

On Wed, Jan 21, 2015 at 6:19 AM, Sean Connolly <notifications@github.com>
wrote:

> I had seen that conversation in #116
> <https://github.com/numenta/htm.java/issues/116> and am still up in the
> air on that one. I'm new to the algorithm and the code base, so can't
> really give a valuable opinion yet.
>
> As for having an SDR class, I think that would be super valuable. Ideally
> it would just be a wrapper of an int[] but would encapsulate what that
> int[] means and does.
>
> If it has no other variables, the footprint overhead would be a few bytes
> just for the Object.
>
> We wouldn't need to worry about instantiation costs if we keep it a simple
> object.
> To this effect I'd shy away from the Tuple implementation we have with
> EncoderResult - that EAV kind of map is probably useful in places, but
> would be wasteful for such a core object.
>
> Think of it like a String; it's a immutable wrapper of a char[] that lets
> you do cool things to it.
>
> As a side bonus: by tucking all this common logic away into one place, it
> would be super easy to do performance tests between different
> implementations. It would be nice to have some benchmarks in place
> beforehand so as to be confident we'd be moving in the right direction with
> this change.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/htm.java/issues/161#issuecomment-70827406>.
>



-- 
*We find it hard to hear what another is saying because of how loudly "who
one is", speaks...*

**Comment by anubhavchaturvedi** [comment]
@cogmission : It would be great if you could explain EncoderResult in a bit more detail. From what I understand EncoderResult just keeps the mapping of the raw input, its numerical representaion if used and the SDR for the input. Now why should we use a Tuple for it just because python did it that way?

From the understanding I have, I think having such a class ( something like SDR.java ) would be effective. Here are my reasons:
* All logic concerning Encoding can be concentrated into a single place. Moreover using boolean[] for encoding in that implementation will automatically fix checks like testing if array contains 0 and 1 only.
* Making the class Immutable will increase performance because of caching. And the SDR will be the most used thing everywhere.

**Comment by cogmission** [comment]
@anubhavchaturvedi I think you are misunderstanding my point because I began the conversation speaking of an Encoding class and an SDR class as if they were to be the same thing (serving the same purpose), but then I realized that they are two distinct propositions with distinct purposes.

We could use both an SDR class *and* we can refine the existing EncoderResult class to more fully encapsulate useful functionality. 

The EncoderResult (which I think should be renamed to "Encoding") is only used as a return value from the topDownCompute method (and maybe other supporting methods), and it is *meant* to contain the input; encoding; and a scalar representation. There are two reasons for this, (one I need to do some further work to verify), 1.) The main requirement of htm.java is to maintain a noticeable one-to-one correspondence with the Python version, which includes the tests so that true equality between the versions can be easily validated. The Python version returns an aggregate which is used in the tests; so I duplicated its functionality.

The SDR class (I've since done some research on this and think this would be beneficial. I'll do a write up on this as a separate discussion); is simply the main product and data transport throughout the entire codebase, so its scope is much broader than EncoderResult, and the two are not the same thing. As a matter of fact, EncoderResult may evolve to contain a formal SDR object rather than an int array if we switch to an SDR object as bit vector representation.

**Comment by cogmission** [comment]
@connollyst @anubhavchaturvedi @solovyevk @subutai @scottpurdy @chetan51 @rhyolight 
*It would be nice to have some benchmarks in place beforehand so as to be confident we'd be moving in the right direction with this change.*

**Done!**

We now have benchmarks! They're added to src/jmh/java <-- for reasons having to do with being able to invoke them from gradle without having to have a separate project. Then can be invoked by the command "gradle check" (or gradle -Pskipbench check to skip), and can be invoked separately by going to the build/libs/jmh directory (after running "gradle check") and invoking the command java -jar <htm-xxx-.jar> 

Now we can look at a true comparison of a Roaring Bitmap or some other compressed bitmap alternative to an integer array and get some true numbers!

I wonder if any of the Numenta engineers have had this discussion (regarding compressed bitmap representation of SDR's?)
